app-ng/src/entry-point.lisp

```diff
@@ -0,0 +1,163 @@
+;;;; app-ng/src/entry-point.lisp
+;;;;
+;;;; Author: appleby
+
+(in-package #:qvm-app-ng)
+
+(defun show-help ()
+  (format t "Usage:~%")
+  (format t "    ~A [<options>...]~%~%" *program-name*)
+  (format t "Options:~%")
+  (command-line-arguments:show-option-help *option-spec* :sort-names t))
+
+(defun show-version ()
+  (format t "~A [~A]~%" +QVM-VERSION+ +GIT-HASH+))
+
+(defun show-welcome ()
+  (format t "~&~
+*****************************************~%~
+* Welcome to the Rigetti QVM (Next Gen) *~%~
+*****************************************~%~
+Copyright (c) 2016-2019 Rigetti Computing.~2%")
+  #+forest-sdk
+  (format t "This is a part of the Forest SDK. By using this program~%~
+             you agree to the End User License Agreement (EULA) supplied~%~
+             with this program. If you did not receive the EULA, please~%~
+             contact <support@rigetti.com>.~2%")
+  (format t "(Configured with ~A MiB of workspace and ~D worker~:P.)~2%"
+          #+sbcl
+          (floor (sb-ext:dynamic-space-size) (expt 1024 2))
+          #-sbcl
+          "many many"
+          (or *num-workers* (max 1 (qvm:count-logical-cores))))
+  nil)
+
+(defun quit-nicely (&optional (code 0))
+  #+sbcl
+  (sb-ext:exit :code code :abort nil)
+  #-sbcl
+  (uiop:quit code t))
+
+(defun process-options (&key
+                          version
+                          check-libraries
+                          verbose
+                          help
+                          #-forest-sdk swank-port
+                          num-workers
+                          #-forest-sdk debug
+                          check-sdk-version
+                          proxy
+                          quiet
+                          log-level
+                          (batch t)
+                          server
+                          simulation-method
+                          allocation-method
+                          memory-limit
+                          safe-include-directory
+                          qubits)
+  "Process all options, setting global variables, and process input or start server."
+  ;; Initialize global configuration first by loading the config file.
+  ;; TODO Specify the file on command line.
+  (load-config-file :reset t)
+  
+  (initialize-logger "qvm-app-ng" log-level)
+
+  (when help
+    (show-help)
+    (quit-nicely))
+
+  (when version
+    (show-version)
+    (quit-nicely))
+
+  (when check-libraries
+    (quit-nicely (if (check-libraries) 0 1)))
+
+  (when check-sdk-version
+    (quit-nicely (if (check-sdk-version :proxy proxy) 0 1)))
+
+  (when verbose
+    (setf qvm:*transition-verbose* t))
+
+  #-forest-sdk
+  (when debug
+    (setf *debug* t))
+
+  (cond
+    ((minusp num-workers)
+     (error "Cannot create a negative number (~D) of workers." num-workers))
+    ((zerop num-workers)
+     (qvm:prepare-for-parallelization)
+     (setf (gethash "num-workers" *config*) 0))
```

Comment:
What happens if `num-workers` was specified in the config file but not on the command line? I think the first thing that happens at the top of `process-options` should be to merge the config file and command line options into `*config*`, then read from the merged `*config*` everywhere inside this function.

Reply:
I also think there is something to be said for having `process-options` only create the merged `*config*`, then have a separate `(initialize-app merged-config)` that does initialization and startup of server/batch mode. I think that would make it easier to test the two in isolation, i.e. tests for config merging vs tests for startup/initialization.